projects/canviewer/canviewer.py

```diff
@@ -0,0 +1,117 @@
+from canserver import init_can
+from gui.gui import Window
+
+import argparse
+import sys
+import time
+import signal
+import numpy as np
+import yaml
+import cantools
+
+from PyQt5.QtWidgets import QApplication
+from PyQt5.QtCore import QTimer
+from rich.layout import Layout
+from rich.live import Live
+
+db = None
+
+with open("projects/canviewer/config.yml", "r") as config_file:
+    (
+        SHUTDOWN_NODES,
+        VEHICLE_VALUES,
+        VEHICLE_STATES,
+    ) = yaml.safe_load_all(config_file)
+# Messages that correspond to vehicle states/faults
+VEHICLE_STATE_SIGNALS = [
+    signal for message in VEHICLE_STATES.values() for signal in message.keys()
+]
+
+
+def sigint(signal_received, frame):
+    exit(0)
+
+
+def convertVtoT(x, Vin=3, R1=10000, R2=100000, T2=348.15, beta=3988):
+    """
+    converts voltage drop data to temperature
+    x: array containing only the voltages
+    """
+    Vout = float(x)
+    thermistor_R1 = (Vout * R1) / (Vin - Vout)
+    temperature = 1 / ((np.log(thermistor_R1 / R2) / beta) + (1 / T2)) - 273.15
+
+    return temperature
+
+
+def get_val(signal, data):
+    val = data.get(signal)
+
+    if val is not None:
+        return str(val)
+
+
+def rx_callback(msg, db):
+    global SHUTDOWN_NODES
+    global VEHICLE_VALUES
+
+    try:
+        message = db.decode_message(msg.arbitration_id, msg.data)
+    except Exception as e:
+        return
+
+    for signal_name in message:
+        if signal_name in SHUTDOWN_NODES:
+            SHUTDOWN_NODES[signal_name] = get_val(signal_name, message)
+        elif signal_name in VEHICLE_VALUES:
+            if signal_name in ["max_temperature", "min_temperature"]:
```

Comment:
I know i wrote this, but i dislike it lmao. it would be cool if there was a way to move the logic for processing a signal elsewhere, like maybe you can put the function name in the YAML file, and then you just see if the signal has a processing function, and if so, you call it. Make sense? I can clarify offline.

Reply:
Right now the signals are key value pairs in the YAML, with the key being the name and the value being whatever the CAN thread gives us. We could
- add a separate document to the YAML with key value pairs of signal names and processing functions if they require one. It's the path of least resistance, and leaves a simple config document, but it does require adding another document to the YAML
- Change the structure of the YAML so the values are no longer lists
```
---
# Vehicle Values
max_temperature: [~, processing_function]
throttle_pos: [~]
```
Then the main processor would have to try to get the element at index 1. If it's there, run it through the appropriate processor, if not, send it straight back to eliminate some of the YAML verbosity.

This would not require an extra document, but it would add in that list into our config, which no longer allows the really simple `signal_name:` configuration.

Thoughts?

Reply:
Why can't you do

```yaml
---
max_temperature:
  processing_function: convert_v_to_t
throttle_pos:
something_else:
```

And then in code you can check to see if the keys have any values, and if so, look for something called `processing_function`?

I am in favor of keeping everything in a single config file.

Reply:
Well it would be in the same config file, just another YAML document. In any case, I'll do some testing but yeah this is a much better solution :+1: